### ISSUE TITLE
Antialiasing not working correctly

### ISSUE THREAD

**Issue by charbelnicolas** [artemsen/swayimg#183]
Hi Artem, I've noticed that the current implementation of the antialiasing doesn't work correctly. I've made a comparison of a thumbnail antialiased by swayimg against imagemagick's cubic filter and there is a massive difference.

I have set my thumbnails to a 300 x 300 pixel size so I ran this command on the images with imagemagick:

`magick image_file.jpg -resize 300x300^ -filter cubic -gravity center -extent 300x300 image_file.png`

![ss_2024_08_29_00_48_47](https://github.com/user-attachments/assets/39852520-bb61-4a64-9014-dee7ef7695a8)

Please check the top row images and compare the left image (swayimg scaling) with the right image (imagemagick scaling). The jagermeister image is inverted, the right image is swayimg's scaling and the left one is imagemagick's scaling.

I think the scaling/antialiasing can be further improved, as it is right now, it is as if nothing is being done to antialias the images. I guess this applies to both viewer and gallery mode although it is more noticeable when viweing the small thumbnails.

**Comment by charbelnicolas** [comment]
Here are the images I used for the screenshot above:

[https://drive.google.com/drive/folders/1uBYhZc1STvh3qZPcyRWNxExgPBrLkpB2?usp=sharing](https://drive.google.com/drive/folders/1uBYhZc1STvh3qZPcyRWNxExgPBrLkpB2?usp=sharing)

**Comment by artemsen** [comment]
Yes, unfortunately the current implementation of antialiasing only works well for zooming.
To be honest, I'm afraid to change it because it's [not very simple](https://github.com/artemsen/swayimg/blob/36949f044da695d550fd84ff58809664dbccc24e/src/pixmap.c#L251) and I've already forgotten what I wrote there =)

**Comment by charbelnicolas** [comment]
I found this website from the creator of the Magic Kernel which is supposedly what facebook and instagram use to scale their images... It even has a 2021 Reference ANSI C implementation of the Magic Kernel Sharp (MIT-0 License).

[https://johncostella.com/magic/](https://johncostella.com/magic/)

[https://johncostella.com/magic/magic-kernel-sharp.tar.gz](https://johncostella.com/magic/magic-kernel-sharp.tar.gz)

Maybe it is worth having a look, I really have no idea how hard it could be to implement it in swayimg

**Comment by artemsen** [comment]
Thank you!
Looks interesting but it will take a lot of time to study this code.
Image transformation is math, but I'm more of an engineer than a mathematician =)

**Comment by charbelnicolas** [comment]
Hello Artem! Sorry for bothering you again. I had the chance to check the antialiasing today and unfortunately I did not notice any improvement, just the jagged edges shifting in place. I've made a comparison between feh's antialiasing and swayimg's and there is a vast difference:

feh aliased:
![ss_2024_09_16_14_07_25](https://github.com/user-attachments/assets/1bcd19b2-fb4e-4180-92df-7bd69033bad0)

feh antialiased:
![ss_2024_09_16_14_07_27](https://github.com/user-attachments/assets/25692765-61fd-44b7-98ec-4cd32d0db178)

swayimg aliased:
![ss_2024_09_16_14_07_34](https://github.com/user-attachments/assets/ed656305-e8be-4787-8510-cef76190c196)

swayimg antialiased:
![ss_2024_09_16_14_07_36](https://github.com/user-attachments/assets/97f799ac-7041-4a52-a536-85358403d8b2)

As a matter of fact it is impossible to distinguish when antialiasing is enabled in swayimg since it doesn't really change that much.

**Comment by abewieland** [comment]
FYI, I've started work on antialiasing; I did quite a bit of research into MKS and its ANSI C reference code, plus the Rust implementation referenced from its website. I plan to implement MKS for swayimg (and maybe take another look at the bicubic scaling); this is the reason I've recently started opening issues and pull requests.

**Comment by charbelnicolas** [comment]
> FYI, I've started work on antialiasing; I did quite a bit of research into MKS and its ANSI C reference code, plus the Rust implementation referenced from its website. I plan to implement MKS for swayimg (and maybe take another look at the bicubic scaling); this is the reason I've recently started opening issues and pull requests.

That's great news!

**Comment by charbelnicolas** [comment]
With the latest commits everything seems to be working pretty well now! Antialiasing is looking pretty good!
